pipelied: UT: failure: test_restart_resilience.py

### Your Environment

- **Version:** master
- **Affected Component:** AGW
- **Affected Subcomponent:** pipelined
- **Deployment Environment:** CICD

### Describe the Issue
```
[magma@10.240.0.29] out: [vagrant@127.0.0.1:2222] out: --------------------- >> end captured logging << ---------------------
[magma@10.240.0.29] out: [vagrant@127.0.0.1:2222] out: 
[magma@10.240.0.29] out: [vagrant@127.0.0.1:2222] out: ======================================================================
[magma@10.240.0.29] out: [vagrant@127.0.0.1:2222] out: FAIL: Partial redirection test, checks if flows were added properly for url
[magma@10.240.0.29] out: [vagrant@127.0.0.1:2222] out: ----------------------------------------------------------------------
[magma@10.240.0.29] out: [vagrant@127.0.0.1:2222] out: Traceback (most recent call last):
[magma@10.240.0.29] out: [vagrant@127.0.0.1:2222] out:   File "/home/vagrant/magma/lte/gateway/python/magma/pipelined/tests/test_restart_resilience.py", line 545, in test_url_redirect
[magma@10.240.0.29] out: [vagrant@127.0.0.1:2222] out:     pkt_sender.send(packet)
[magma@10.240.0.29] out: [vagrant@127.0.0.1:2222] out:   File "/home/vagrant/magma/lte/gateway/python/magma/pipelined/tests/pipelined_test_util.py", line 589, in __exit__
[magma@10.240.0.29] out: [vagrant@127.0.0.1:2222] out:     self._snapshot_name, self._include_stats)
[magma@10.240.0.29] out: [vagrant@127.0.0.1:2222] out:   File "/home/vagrant/magma/lte/gateway/python/magma/pipelined/tests/pipelined_test_util.py", line 482, in assert_bridge_snapshot_match
[magma@10.240.0.29] out: [vagrant@127.0.0.1:2222] out:     current_snapshot)
[magma@10.240.0.29] out: [vagrant@127.0.0.1:2222] out:   File "/home/vagrant/magma/lte/gateway/python/magma/pipelined/tests/pipelined_test_util.py", line 422, in fail
[magma@10.240.0.29] out: [vagrant@127.0.0.1:2222] out:     return test_case.fail(msg)
[magma@10.240.0.29] out: [vagrant@127.0.0.1:2222] out: AssertionError: Snapshot mismatch with error:
[magma@10.240.0.29] out: [vagrant@127.0.0.1:2222] out: --- previous snapshot
[magma@10.240.0.29] out: [vagrant@127.0.0.1:2222] out: 
[magma@10.240.0.29] out: [vagrant@127.0.0.1:2222] out: +++ current snapshot
```

**Additional context**
https://circleci.com/api/v1.1/project/github/magma/magma/92241/output/110/0?file=true&allocation-id=5fd170aba209b93597127b1d-0-build%2F1CE7EB96

--- a/lte/gateway/python/magma/pipelined/app/enforcement.py
+++ b/lte/gateway/python/magma/pipelined/app/enforcement.py
@@ -33,6 +33,7 @@ from ryu.controller import ofp_event
 from ryu.controller.handler import MAIN_DISPATCHER, set_ev_cls
 from ryu.ofproto.ofproto_v1_4_parser import OFPFlowStats
 from magma.pipelined.utils import Utils
+from magma.pipelined.openflow.exceptions import MagmaDPDisconnectedError
 
 
 class EnforcementController(PolicyMixin, MagmaController):
@@ -222,7 +223,12 @@ class EnforcementController(PolicyMixin, MagmaController):
         except FlowMatchError:
             return RuleModResult.FAILURE
 
-        chan = self._msg_hub.send(flow_adds, self._datapath)
+        try:
+            chan = self._msg_hub.send(flow_adds, self._datapath)
+        except MagmaDPDisconnectedError:
+            self.logger.error("Datapath disconnected, failed to install rule %s"
+                              "for imsi %s", rule, imsi)
+            return RuleModResult.FAILURE
         return self._wait_for_rule_responses(imsi, ip_addr, rule, chan)
 
     def _install_redirect_flow(self, imsi, ip_addr, rule):

--- a/lte/gateway/python/magma/pipelined/app/enforcement_stats.py
+++ b/lte/gateway/python/magma/pipelined/app/enforcement_stats.py
@@ -39,6 +39,7 @@ from magma.pipelined.openflow.messages import MsgChannel, MessageHub
 from magma.pipelined.utils import Utils
 from magma.pipelined.openflow.registers import Direction, DIRECTION_REG, \
     IMSI_REG, RULE_VERSION_REG, SCRATCH_REGS
+from magma.pipelined.openflow.exceptions import MagmaDPDisconnectedError
 
 
 ETH_FRAME_SIZE_BYTES = 14
@@ -168,7 +169,12 @@ class EnforcementStatsController(PolicyMixin, MagmaController):
 
         msgs = self._get_rule_match_flow_msgs(imsi, msisdn, uplink_tunnel, ip_addr, apn_ambr, rule)
 
-        chan = self._msg_hub.send(msgs, self._datapath)
+        try:
+            chan = self._msg_hub.send(msgs, self._datapath)
+        except MagmaDPDisconnectedError:
+            self.logger.error("Datapath disconnected, failed to install rule %s"
+                              "for imsi %s", rule, imsi)
+            return RuleModResult.FAILURE
         for _ in range(len(msgs)):
             try:
                 result = chan.get()

--- a/lte/gateway/python/magma/pipelined/app/ipv6_solicitation.py
+++ b/lte/gateway/python/magma/pipelined/app/ipv6_solicitation.py
@@ -193,7 +193,9 @@ class IPV6SolicitationController(MagmaController):
                                   in_port=ofproto.OFPP_CONTROLLER,
                                   actions=actions_out,
                                   data=pkt.data)
-        self._datapath.send_msg(out)
+        ret = self._datapath.send_msg(out)
+        if not ret:
+            self.logger.error("Datapath disconnected, couldn't send RA")
 
     def _send_neighbor_advertisement(self, target_ipv6: str, tun_id: int,
                                      tun_ipv4_dst: str, output_port, direction):
@@ -251,7 +253,9 @@ class IPV6SolicitationController(MagmaController):
                                   in_port=ofproto.OFPP_CONTROLLER,
                                   actions=actions_out,
                                   data=pkt.data)
-        self._datapath.send_msg(out)
+        ret = self._datapath.send_msg(out)
+        if not ret:
+            self.logger.error("Datapath disconnected, couldn't send NA")
 
     @set_ev_cls(ofp_event.EventOFPPacketIn, MAIN_DISPATCHER)
     def _parse_pkt_in(self, ev):

--- a/lte/gateway/python/magma/pipelined/openflow/exceptions.py
+++ b/lte/gateway/python/magma/pipelined/openflow/exceptions.py
@@ -15,3 +15,9 @@ class MagmaOFError(Exception):
     Base class for OpenFlow-related exceptions.
     """
     pass
+
+class MagmaDPDisconnectedError(Exception):
+    """
+    Exception class for ryu Datapath disconnect.
+    """
+    pass

--- a/lte/gateway/python/magma/pipelined/openflow/messages.py
+++ b/lte/gateway/python/magma/pipelined/openflow/messages.py
@@ -19,7 +19,8 @@ from ryu.controller.controller import Datapath
 from ryu.lib import hub
 from ryu.ofproto.ofproto_parser import MsgBase
 
-from magma.pipelined.openflow.exceptions import MagmaOFError
+from magma.pipelined.openflow.exceptions import MagmaOFError,\
+    MagmaDPDisconnectedError
 from magma.pipelined.metrics import DP_SEND_MSG_ERROR
 
 logger = logging.getLogger(__name__)
@@ -45,7 +46,9 @@ def send_msg(datapath, msg, retries=3):
     """
     for i in range(0, retries):
         try:
-            datapath.send_msg(msg)
+            ret = datapath.send_msg(msg)
+            if not ret:
+                raise MagmaDPDisconnectedError()
             return
         except Exception as e:  # pylint: disable=broad-except
             logger.warning(
@@ -165,7 +168,9 @@ class MessageHub(object):
         switch.requests_by_barrier[barrier.xid] = req
         for msg in msg_list:
             switch.results_by_msg[msg.xid] = None
-            datapath.send_msg(msg)
+            ret = datapath.send_msg(msg)
+            if not ret:
+                raise MagmaDPDisconnectedError()
         datapath.send_msg(barrier)
         req.set_timeout(timeout, switch, barrier.xid, msg_xids)
         return channel

